### DISCUSSION TITLE
Thoughts on API design and generation

### DISCUSSION THREAD

**Discussion by fserb** [progrium/darwinkit#73] [Ideas]
Given that I just learned about MacDriver and how to use it 3 days ago, I thought it would be nice to ask some questions about the design of the overall package.

I looked around PyObjC and RubyCocoa and kept wondering how hard it would be to do something similar to MacDriver. Have an automated way to generate some of the bindings based on Headers? Is this something worth trying? I wanted to take a look at that, maybe starting with enums/consts? Ideas? Suggestions?

Is there any reason why the constructors were settled on `cocoa.NSWhatever_Init()`, as opposed to `cocoa.NSWhatever.New()` or `cocoa.NSWhatever.Init()`? I get it that, because of go module/directory thing, you'd end up with one directory per class, which is mildly annoying. But the final API would look nicer? Maybe this would be less of a problem if it was auto generated?

I see that most classes are just a bunch of getters/setters. They mostly don't include any extra sugar, like for delegation, events, blocks, etc. It would be nice to have some generic ways to handle those (I'm thinking about `NSResponder` for example).

anyways, just wanted to drop some notes here. I'll keep digging around. :)

**Comment by progrium** [comment]
We've started using generation, but its a work in progress. We can't do `cocoa.NSWhatever.New()` because then we'd have to come up with a different name for the type `NSWhatever`. Delegation and even subclassing works with Go created classes now. Blocks I don't imagine work though. Thanks for the suggestions.
